wren-ui/src/apollo/server/services/dashboardService.ts

```diff
@@ -193,4 +254,270 @@ export class DashboardService implements IDashboardService {
+        schedule.frequency,
+      ) ||
+      !timezone
+    ) {
+      return schedule;
+    }
+
+    // Get timezone offset in minutes
+    const offsetMinutes = getUTCOffsetMinutes(timezone);
+
+    // Convert from UTC to timezone by adding the offset (if timezone is ahead of UTC)
+    // or subtracting the offset (if timezone is behind UTC)
+    let localMinute = schedule.minute + offsetMinutes;
+    let carryOver = 0;
+
+    // Handle minute carry-over
+    if (localMinute < 0) {
+      carryOver = Math.ceil(Math.abs(localMinute) / 60);
+      localMinute = (localMinute + carryOver * 60) % 60;
+      carryOver = -carryOver;
+    } else if (localMinute >= 60) {
+      carryOver = Math.floor(localMinute / 60);
+      localMinute = localMinute % 60;
+    }
+
+    let localHour = schedule.hour + carryOver;
+    let dayAdjustment = 0;
+
+    // Handle hour carry-over
+    if (localHour < 0) {
+      dayAdjustment = Math.ceil(Math.abs(localHour) / 24);
+      localHour = (localHour + dayAdjustment * 24) % 24;
+      dayAdjustment = -dayAdjustment;
+    } else if (localHour >= 24) {
+      dayAdjustment = Math.floor(localHour / 24);
+      localHour = localHour % 24;
+    }
+
+    // For weekly schedules, adjust the day if needed
+    if (
+      schedule.frequency === ScheduleFrequencyEnum.WEEKLY &&
+      dayAdjustment !== 0
+    ) {
+      const currentDayIndex = DAYS.indexOf(schedule.day);
+      const adjustedDayIndex = (currentDayIndex + dayAdjustment + 7) % 7;
+      return {
+        ...schedule,
+        hour: localHour,
+        minute: localMinute,
+        day: DAYS[adjustedDayIndex],
+        timezone,
+      };
+    }
+
+    // Return a new schedule object with local hours and minutes
+    return {
+      ...schedule,
+      hour: localHour,
+      minute: localMinute,
+      timezone,
+    };
+  }
+
+  protected generateCronExpression(schedule: DashboardSchedule): string | null {
+    const { frequency, day, hour, minute } = this.toUTC(schedule);
+
+    switch (frequency) {
+      case ScheduleFrequencyEnum.DAILY:
+        return `${minute} ${hour} * * *`;
+      case ScheduleFrequencyEnum.WEEKLY:
+        // e.g. 0 10 * * MON
+        return `${minute} ${hour} * * ${day}`;
+      case ScheduleFrequencyEnum.CUSTOM:
+        return schedule.cron;
+      case ScheduleFrequencyEnum.NEVER:
+        return null;
+      default:
+        logger.warn(`Unsupported schedule frequency: ${frequency}`);
+        return null;
+    }
+  }
+
+  protected calculateNextRunTime(cronExpression: string): Date | null {
+    try {
+      const interval = CronExpressionParser.parse(cronExpression, {
+        currentDate: new Date(),
+      });
+      return interval.next().toDate();
+    } catch (error) {
+      logger.error(`Failed to parse cron expression: ${error.message}`);
+      return null;
+    }
+  }
+
+  protected validateScheduleInput(data: SetDashboardCacheData): void {
```

Comment:
it's actually could be `private` scope, correct ? seems not used elsewhere besides this file.

Reply:
I see, you extend this service and use it in testing. Well, maybe putting in a util file to test would be easier ?

Reply:
Just a nitpick. It should be ok for now.